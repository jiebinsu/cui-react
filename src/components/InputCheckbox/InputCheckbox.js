import React, { useState, useEffect, useCallback, forwardRef } from "react";
import PropTypes from "prop-types";
import { useCheckboxContext } from "../InputCheckboxGroup/InputCheckboxGroupContext";
import classNames from "../../utils/classNames";
import "./inputCheckbox.scss";

const InputCheckbox = forwardRef(
  (
    {
      id,
      name,
      className,
      label,
      checked,
      value: checkboxValue,
      disabled,
      onChange,
      ...props
    },
    ref
  ) => {
    const [selfChecked, setSelfChecked] = useState(!!checked);
    const { updateState, inGroup, values } = useCheckboxContext();
    const classes = classNames(
      "coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox",
      [className]
    );

    if (inGroup) {
      useEffect(() => {
        const next = values.includes(checkboxValue);
        if (next === selfChecked) return;
        setSelfChecked(next);
      }, [values.join(",")]);
    }

    const onChangeHandler = useCallback(
      (event) => {
        if (disabled) return;

        const selfEvent = {
          target: {
            checked: !selfChecked,
          },
          stopPropagation: event.stopPropagation,
          preventDefault: event.preventDefault,
          nativeEvent: event,
        };
        if (inGroup && updateState) {
          if (updateState) updateState(checkboxValue, !selfChecked);
        }

        setSelfChecked(!selfChecked);
        if (onChange) onChange(selfEvent);
      },
      [updateState, onChange, disabled, selfChecked]
    );

    useEffect(() => {
      if (checked === undefined) return;
      setSelfChecked(!!checked);
    }, [checked]);

    const tagAttributes = {
      id,
      name,
      value: checkboxValue,
      label,
      disabled,
      type: "checkbox",
      className: classes,
      checked: selfChecked,
      onChange: onChangeHandler,
      ...props,
      ref,
    };

    return (
      <div className="coop-c-form-choice">
        <input {...tagAttributes} />
        <label
          className="coop-form__label coop-c-form-choice__label"
          htmlFor={id}
        >
          {label}
        </label>
      </div>
    );
  }
);

InputCheckbox.defaultProps = {
  disabled: false,
  className: null,
  checked: null,
  onChange: null,
  value: "",
};

InputCheckbox.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  className: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

export default InputCheckbox;
