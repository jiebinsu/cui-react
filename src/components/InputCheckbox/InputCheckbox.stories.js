import React from "react";
import InputCheckbox from "./InputCheckbox";

export default {
  title: "Base/InputCheckbox",
  component: InputCheckbox,
};

export const Basic = (args) => <InputCheckbox {...args} />;
Basic.args = {
  id: "basic",
  name: "basic",
  label: "Basic Checkbox",
  value: 1,
};
