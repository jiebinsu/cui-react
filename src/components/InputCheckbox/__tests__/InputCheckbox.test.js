import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import InputCheckbox from "../index";

describe("InputCheckbox", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <InputCheckbox id="test" name="test" label="test" value={1} />
    );
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer
      .create(<InputCheckbox id="test" name="test" label="test" value={1} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
