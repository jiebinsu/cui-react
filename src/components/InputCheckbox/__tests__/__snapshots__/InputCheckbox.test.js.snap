// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCheckbox should render correctly 1`] = `
<div
  className="coop-c-form-choice"
>
  <input
    checked={false}
    className="coop-form__field coop-form__checkbox coop-c-form-choice__input coop-c-form-choice__input--checkbox"
    disabled={false}
    id="test"
    label="test"
    name="test"
    onChange={[Function]}
    type="checkbox"
    value={1}
  />
  <label
    className="coop-form__label coop-c-form-choice__label"
    htmlFor="test"
  >
    test
  </label>
</div>
`;
